lib/inputstreamhelper/__init__.py

```diff
@@ -540,6 +551,45 @@ def _chromeos_config(self):
 
         return devices
 
+    def _remove_old_backups(self, backup_path):
+        """Removes old Widevine backups, if number of allowed backups is exceeded"""
+        max_backups = int(get_setting('backups', '4'))
+        versions = sorted([LooseVersion(version) for version in os.listdir(backup_path)])
+
+        if len(versions) < 2:
+            return
+
+        if 'x86' in self._arch():
+            installed_version = self._load_widevine_config()['version']
+        else:
+            installed_version = self._select_best_chromeos_image(self._load_widevine_config())['version']
+
+        while len(versions) > max_backups + 1:
+            remove_version = str(versions[1] if versions[0] == installed_version else versions[0])
+            log('removing oldest backup which is not installed: {version}', version=remove_version)
+            shutil.rmtree(os.path.join(backup_path, remove_version))
+            versions = sorted([LooseVersion(version) for version in os.listdir(backup_path)])
+
+        return
+
+    def _install_cdm_from_backup(self, version):
+        """Copies files from specified backup version to cdm dir"""
+        filenames = os.listdir(os.path.join(self._backup_path(), version))
+
+        for filename in filenames:
+            backup_fpath = os.path.join(self._backup_path(), version, filename)
+            install_fpath = os.path.join(self._ia_cdm_path(), filename)
+
+            if system_os() == 'Windows' and version_info[0] < 3:
+                xbmcvfs.copy(backup_fpath, install_fpath)
+            else:
+                if xbmcvfs.exists(install_fpath):
+                    xbmcvfs.delete(install_fpath)
+                os.link(backup_fpath, install_fpath)
```

Comment:
Two comments:
- `os.link()` shouldn't require a delete, else it would not be atomic
- I think you can simply try `os.link()` and when you get an OSError exception perform a copy

This should work:
```python
            try:
                os.link(backup_fpath, install_fpath)
            except OSError:
                xbmcvfs.copy(backup_fpath, install_fpath)
```

Reply:
`os.link()` gives an OSError on Python2 and a FileExistsError on python3 if the destination exists. I had checked that. But I'll change it to try...except.

Reply:
This is truly unfortunate that anno 2019 there is no race-free implementation of `os.link()`.
https://bugs.python.org/issue36656

Revert to what you have, I'll crawl under a stone.